File: src/main/java/com/chattriggers/ctjs/minecraft/objects/display/DisplayHandler.java
package com.chattriggers.ctjs.minecraft.objects.display;

import com.chattriggers.ctjs.minecraft.libs.EventLib;
import lombok.Getter;
import net.minecraft.client.renderer.GlStateManager;
import net.minecraftforge.client.event.RenderGameOverlayEvent;
import net.minecraftforge.common.MinecraftForge;
import net.minecraftforge.fml.common.eventhandler.SubscribeEvent;

import java.util.ArrayList;

public class DisplayHandler {
    private static DisplayHandler instance;

    public static DisplayHandler getInstance() {
        return instance;
    }

    private ArrayList<Display> displays;

    public DisplayHandler() {
        instance = this;
        MinecraftForge.EVENT_BUS.register(this);

        displays = new ArrayList<>();
    }

    public void registerDisplay(Display display) {
        this.displays.add(display);
    }

    public void clearDisplays() {
        displays.clear();
    }

    @SubscribeEvent
    public void renderDisplays(RenderGameOverlayEvent event) {
        GlStateManager.pushMatrix();

        if (EventLib.getType(event) == RenderGameOverlayEvent.ElementType.TEXT) {
            displays.forEach(Display::render);
        }

        GlStateManager.popMatrix();
    }

    /**
     * The background type of a display.
     */
    public enum Background {
        NONE, FULL, PER_LINE;

        public static Background getBackgroundByName(String background) {
            return Background.valueOf(background.toUpperCase().replace(" ", "_"));
        }
    }

    /**
     * The alignment of the text in a display.
     */
    public enum Align {
        NONE, LEFT, CENTER, RIGHT;

        public static Align getAlignByName(String align) {
            return Align.valueOf(align.toUpperCase());
        }
    }

    /**
     * The order that a display should be drawn in.
     */
    public enum Order {
        UP, DOWN;

        public static Order getOrderByName(String order) {
            return Order.valueOf(order.toUpperCase());
        }
    }
}

File: src/main/kotlin/com/chattriggers/ctjs/minecraft/objects/display/DisplayHandler.kt
package com.chattriggers.ctjs.minecraft.objects.display

import com.chattriggers.ctjs.CTJS
import com.chattriggers.ctjs.minecraft.libs.renderer.Renderer
import com.chattriggers.ctjs.triggers.EventType
import java.util.concurrent.CopyOnWriteArrayList

object DisplayHandler {
    private val displays = CopyOnWriteArrayList<Display>()

    init {
        CTJS.addEventListener(EventType.RenderOverlay) {
            Renderer.pushMatrix()
            displays.forEach {
                if (it.registerType == RegisterType.RENDER_OVERLAY) {
                    it.render()
                }
            }
            Renderer.popMatrix()
        }

        CTJS.addEventListener(EventType.PostGuiRender) {
            Renderer.pushMatrix()
            displays.forEach {
                if (it.registerType == RegisterType.POST_GUI_RENDER) {
                    it.render()
                }
            }
            Renderer.popMatrix()
        }
    }

    fun registerDisplay(display: Display) = displays.add(display)

    fun clearDisplays() = displays.clear()

    enum class RegisterType {
        RENDER_OVERLAY, POST_GUI_RENDER
    }

    enum class Background {
        NONE, FULL, PER_LINE
    }

    enum class Align {
        LEFT, CENTER, RIGHT
    }

    enum class Order {
        UP, DOWN
    }
}
